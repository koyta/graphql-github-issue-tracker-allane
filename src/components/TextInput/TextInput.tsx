import { ChangeEvent, ComponentPropsWithoutRef, forwardRef } from "react";

import classNames from "classnames";

type Props = {
  icon?: JSX.Element;
  onChange?: (value: string) => void;
} & Omit<ComponentPropsWithoutRef<"input">, "onChange">;

export const TextInput = forwardRef<HTMLInputElement, Props>(
  ({ onChange, value, className, icon, ...props }, ref) => {
    const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
      if (props.disabled) {
        return;
      }
      onChange?.(event.target.value);
    };

    return (
      <div className="flex flex-row flex-grow items-center relative w-min">
        <input
          ref={ref}
          type="text"
          onChange={handleChange}
          value={value}
          className={classNames(
            "py-2 px-3 pr-7 border border-gray-200 rounded-md",
            className
          )}
          {...props}
        />
        {icon && <i className="absolute right-2">{icon}</i>}
      </div>
    );
  }
);
