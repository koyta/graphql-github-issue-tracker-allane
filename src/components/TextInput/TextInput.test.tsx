import { fireEvent, render } from "@testing-library/react";

import { TextInput } from "@app/components/TextInput/TextInput";

describe("TextInput", () => {
  it("should render a text input", () => {
    const { container } = render(<TextInput />);
    expect(container).toMatchSnapshot();
  });
  it("should be controlled", () => {
    const { container } = render(<TextInput value="Hello" />);
    const input = container.querySelector("input");
    if (!input) {
      throw new Error("Input not found");
    }
    expect(input).toHaveValue("Hello");
  });
  it("should call onChange callback on change", () => {
    const onChange = vi.fn();
    const { container } = render(<TextInput onChange={onChange} />);
    const input = container.querySelector("input");
    if (!input) {
      throw new Error("Input not found");
    }
    fireEvent.change(input, { target: { value: "Hello" } });
    expect(onChange).toHaveBeenCalledWith("Hello");
  });
  it("should be disabled", () => {
    const onChange = vi.fn();
    const { container } = render(<TextInput disabled onChange={onChange} />);
    const input = container.querySelector("input");
    if (!input) {
      throw new Error("Input not found");
    }
    fireEvent.change(input, { target: { value: "Hello" } });
    expect(onChange).not.toHaveBeenCalled();
  });
  it("should render a text input with a custom class", () => {
    const { container } = render(<TextInput className="custom-class" />);
    expect(container).toMatchSnapshot();
  });
});
