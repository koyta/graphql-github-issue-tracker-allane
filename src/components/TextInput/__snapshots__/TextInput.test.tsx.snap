// Vitest Snapshot v1

exports[`TextInput > should render a text input 1`] = `
<div>
  <div
    class="flex flex-row items-center relative w-min"
  >
    <input
      class="py-2 px-3 pr-7 border border-gray-200 rounded-md"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`TextInput > should render a text input with a custom class 1`] = `
<div>
  <div
    class="flex flex-row items-center relative w-min"
  >
    <input
      class="py-2 px-3 pr-7 border border-gray-200 rounded-md custom-class"
      type="text"
      value=""
    />
  </div>
</div>
`;
